keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -409,11 +416,198 @@ enhancement:
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+
+The metrics endpoint should consume a tiny fraction of the CPU of the scheduler (less than 5% at idle) when scraped
+every 15s. The endpoint should return quickly (tens of milliseconds at a P99) when O(pods) is below 10,000. CPU and
+latency should be proportional to number of pods only, as the rest of the scheduler, and the metrics endpoint should
+scale linearly to that factor.
+
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+
+No
+
+### Dependencies
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **Does this feature depend on any specific services running in the cluster?**
+
+  - Scheduler
+    - Hosts the metrics
+  - Metrics collector
+    - Scrapes the endpoint
+    - May run on or off clutser
+
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+
+No, this pulls directly from the scheduler's informer cache.
+
+* **Will enabling / using this feature result in introducing new API types?**
+
+No.
+
+* **Will enabling / using this feature result in any new calls to the cloud
+provider?**
+
+No.
+
+* **Will enabling / using this feature result in increasing size or count of
+the existing API objects?**
+
+No.
+
+* **Will enabling / using this feature result in increasing time taken by any
+operations covered by [existing SLIs/SLOs]?**
+
+The CPU usage of this feature when activated should have a negligible effect on
+scheduler throughput and latency. No additional memory usage is expected.
+
+* **Will enabling / using this feature result in non-negligible increase of
+resource usage (CPU, RAM, disk, IO, ...) in any components?**
+
+Negligible CPU use is expected and some increase in network transmit when the scheduler
+is scraped.
+
+### Troubleshooting
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+It returns the metrics of the last set of data received by the scheduler, or no
+metrics if the scheduler has been restarted since partitioned from the API server.
+
+* **What are other known failure modes?**
+
+  - Panic due to unexpected code path or incomplete API objects returned in watch
+    - Detection: The scrape of the component should fail
+    - Mitigations: Stop scraping the endpoint
+    - Diagnostics: Panic messages in the scheduler logs
+    - Testing: We do not inject fake panics because the behavior of metrics endpoints are well known and there is no background processing.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+Perform a golang CPU profile of the scheduler and assess the percentage of CPU charged to the functions
```

Comment:
Is there any way (from metrics or logs) to tell how often the metrics endpoints are being scraped?

Reply:
The prometheus default metrics handler should report it (to the scraper), it would show up in a profile, but it does not write a log line (kind of in keeping with our expectations around kube-apiserver).

Reply:
Ok, so if there is a metric from the default metrics handler please list it. This is for operators, so your instructions are more for someone developing it. I think the actions here would be:
- check the metric to see if it's being scraped like crazy
- if so, disable the scraper
- last resort, disable the http server for scheduler metrics, but that has obvious bad consequences for visibility

Reply:
Turns out we don't have those metrics on for request count.  Have we tried to impose a general statement about that?  We're kind of inconsistent but i'd expect sig-instrumentation to blanket enforce that (components that expose metrics should also expose the standard prometheus http request handler).  There are some cardinality concerns.  For right now we'd have to add something to catch that, and we don't *generally* do it.  Would you prefer me to provide a simpler path, or can we punt this to a broader "let's have better standard operational metrics for counting requests that don't blow out cardinality"?

Reply:
we can punt it, maybe create an issue for sig-instrumentation?